### ISSUE TITLE
Building collapse for use with webr

### ISSUE THREAD

**Issue by psychemedia** [r-wasm/jupyterlite-webr-kernel#9]
I've been stuck on this for a couple of hours now, and don't see what I'm doing wrong?

I'm trying to use the `plm` package, which has a dependency `collapse` which pulls in an item in its `NAMESPACE` file from `parallel`, which is understandably not in the core `webr` package. The function that is loaded from `parallel` is guarded by a *parallel=* flag with a default `FALSE`, so I assume that commenting out the `NAMESPACE` item in the padkage shouldn't have a side effect if R doesn't statically check that all functions are declared.

I tried downloading the `collapse` package source, commenting the unwanted import out, tarring the package and then attempting to compile it using the webr package builder action, but couldn't find a way of naming the local file that didn't cause an error?

I noticed an issue on the `collapse` repo (https://github.com/SebKrantz/collapse/issues/550) pointing to what looks like a [non-parallel branch](https://github.com/SebKrantz/collapse/tree/no_parallel_simd), and thought to import that using the package stirng `github::SebKrantz/collapse@no_parallel_simd`. The [build job runs](https://github.com/ouseful-testing/webr-package-repo-demo/actions/runs/10912530005/job/30287362250) but the `collapse` package is not built.

So what am I doing wrong?

**Comment by SebKrantz** [comment]
The no_parallel_simd branch is something else, having to do with OpenMP. The *parallel* package is part of base R now and should be included in *any* R distribution. I only conditionally use the mclapply() function from it. 

Please try the following: remove the import line in the NAMESPACE and replace all instances of mclappy with parallel::mclapply. If that works I can change the code towards this end.

**Comment by psychemedia** [comment]
Thanks, will give that a try.

**Comment by georgestagg** [comment]
I am not sure why `collapse_1.9.6.tar.gz` has not been found in the GitHub Actions run that you have linked, I can see the file exists in the repo. Perhaps something has gone wrong and the archive is not a valid R package anymore.

In any case there are some bigger issues here.

@SebKrantz is correct, the parallel package should be available as part of the webR distribution. You should be able to run `library(parallel)` to load it from the network. It succeeds on https://webr.r-wasm.org/v0.3.3/, for example.

This indicates to me that your base webR distribution is corrupt in some way. IIRC you are using your own `baseUrl`? On your server there should be a directory `vfs` served as part of the base webR distribution. Is that directory available? In there the parallel package is served as under the path `.../vfs/usr/lib/R/library/parallel.data`.

Once that is working, you should be able to load the `collapse` package without modifying its source code at all. Again, as an example, it can be loaded at https://webr.r-wasm.org/v0.3.3/.

----

A bigger problem is that I notice that separately to this you are trying to compile the `curl` package. Unfortunately, that isn't going to work at all.

The curl package relies on the system dependency `libcurl` and we haven't compiled a version of `libcurl` for webR/Wasm yet. In fact, this is a very hard problem because WebAssembly in a browser does not have direct access to the network sockets on the host machine.

The only real way forward is to create a "shim" curl package or system library that replaces HTTP curl requests with JavaScript browser API calls. This is something we're planning on looking at at some point in the future, but the project will be a large undertaking. Curl has a wide API and supports many network protocols, and so it is not entirely clear which subset of curl APIs can be translated into browser APIs, not to speak about actually implementing them.

**Comment by psychemedia** [comment]
`parallel` should be available - my mistake in the webr distribution

**Comment by georgestagg** [comment]
> Re: the tar.gz: I downloaded it from https://cran.r-project.org/src/contrib/Archive/collapse/, made the edit, then tar -czvf collapse_2.0.15.tar.gz collapse (on a Mac).

Interesting, I believe that should work even if the work is not actually required here. It would be interesting to know if the original `.tar.gz` worked. I can see you're compiling other source files using `local::` and they seem to work OK?

**Comment by psychemedia** [comment]
@georgestagg Yes, the others work fine (`MASS` was from the same CRAN source but untouched).
